doc/guides/naming_conventions/index.rst

```diff
@@ -0,0 +1,88 @@
+Naming Conventions
+##################
+
+Functions and Procedures
+************************
+
+#. No procedure shall have a name that is a keyword of any standard version of
+   the C or C++ programming language. Restricted names include ``interrupt``, ``inline``,
+   ``class``, ``true``, ``false``, ``public``, ``private``, ``friend``,
+   ``protected``, and many others.
+
+#. Function names are always in lowercase, with an underscore between each part of
+   the name : ``k_timer_init()`` , ``k_timer_start()``.
+
+#. If a function operates primarily on a certain type, the prefix of this function
+   corresponds to the type name. For example, the k_timer_* functions work with the
+   ``k_timer`` type, and k_sem_* functions go with ``k_sem``.
+
+#. Each procedure’s name shall be descriptive of its purpose. Note that procedures
+   encapsulate the “actions” of a program and thus benefit from the use of verbs in
+   their names (e.g., ``k_mutex_lock()``); this “noun-verb” word ordering is recommended.
+   Alternatively, procedures may be named according to the question they answer
+   (e.g., ``device_is_ready()``).
+
+#. The names of all public functions shall be prefixed with their module name and
+   an underscore (e.g., ``sensor_read()``).
+
+#. Avoid names that differ only in case, like foo and Foo. Similarly, avoid foobar
+   and foo_bar. The potential for confusion is considerable.
+
+#. Similarly, avoid names that look like each other. On many terminals and
+   printers, 'l', '1' and 'I' look quite similar. A variable named 'l' is
+   particularly bad because it looks so much like the constant '1'.
+
+#. #define constants should be in all CAPS.
+
+#. No procedure shall have a name that overlaps a function in the C Standard
+   Library. Examples of such names include strlen, atoi, and memset.
+
+#. No procedure shall have a name that begins with an underscore.
+
+#. No procedure name shall be longer than 31 characters.
+
+#. Variable names, structure tag names, and function names should be in lower
+   case.
+
+#. A prototype shall be declared for each public function in the module header
+   file.
+
+#. All private functions shall be declared ``static``.
+
+#. Each parameter of a function shall be explicitly declared and meaningfully
+   named.
+
+#. Parameterized macros shall not be used if a function can be written to
+   accomplish the same behavior.
+
+#. If parameterized macros are used for some reason, these rules apply:
+
+    #. Surround the entire macro body with parentheses.
+    #. Surround each use of a parameter with parentheses.
+    #. Use each parameter no more than once, to avoid unintended side effects.
```

Comment:
Won't work for various devicetree-style macros.

For example, macros like this have to reuse the same `node_id` in multiple places in their expansion:
https://github.com/zephyrproject-rtos/zephyr/blob/03a1d04c8492beff7318c6003f44cb3f10a34cb7/include/net/net_if.h#L2271

There's no risk of a side-effect here because these sorts of arguments are just tokens that are not statements or expressions.